feat: support extend rules

yi-Xu-0100 opened the pull request:
Add `commitlint.config.extend.rules` configuration to extend rules.

ref: #78 
close: #73
close: #82

joshbolduc commented on src/settings.ts:
```diff
@@ -6,3 +6,10 @@ export function getConfigFile() {
   );
 }
+
+export function getExtendConfiguration(config: string) {
+  return (
+    workspace.getConfiguration('commitlint.config.extend').get(config) ||
+    undefined
+  );
+}
```
I think a generic here could improve the typings:
```suggestion
export function getExtendConfiguration<T extends keyof UserConfig>(config: T) {
  return (
    workspace
      .getConfiguration('commitlint.config.extend')
      .get<UserConfig[T]>(config) || undefined
  );
}
```

Then the type assertion wouldn't be necessary where `getExtendConfiguration` gets called.

joshbolduc commented on src/config.ts:
```diff
@@ -13,5 +16,13 @@ export function loadConfig(path: string | undefined) {
     : { cwd: path };
 
-  return load({}, loadOptions);
+  const config = await load({}, loadOptions);
+
+  for (const rule in extendsRules) {
+    if (!config.rules?.[rule]) {
+      config.rules[rule] = extendsRules[rule];
+    }
+  }
+
+  return config;
```
Could we pass the rules to `load` directly?

```suggestion
  return load({ rules: extendsRules || {} }, loadOptions);
```

yi-Xu-0100 replied:
@joshbolduc I test this before, but it will overwrite the extended configuration setting in the workspace `commitlint.config.js`.

When the workspace `commitlint.config.js` is:
```js
module.exports = {
  extends: ['@commitlint/config-conventional'],
};
```

and the rules I was setting:
```json
"commitlint.config.extend.rules": {
  "body-leading-blank": [3, "always"],
}
```

It will load the rule of "body-leading-blank" with `[3, "always"]`. I think rules should be consistent with the workspace setting.
 And it could add rules which not include in the workspace setting. So I use the loop above to judge whether it should include. 😀

joshbolduc replied:
Makes sense! In that case, I think this might read a little more clearly, and should work the same way?

```ts
const config = await load({}, loadOptions);

return { ...config, rules: { ...extendsRules, ...config.rules } };
```

joshbolduc commented on README.md:
```diff
@@ -11,15 +11,42 @@ A VS Code extension that integrates [commitlint](https://github.com/conventional
 
 ## Settings
 
 ### `commitlint.config.file`
 
 Path to a custom commitlint configuration file. Relative paths are resolved based on the workspace root. Leave blank to auto-detect.
 
+### `commitlint.config.extend.rules`
+
+Commitlint rules which will be extended.
+
+**Example configuration in JSON type:**
```
```suggestion
### `commitlint.config.file`

Path to a custom commitlint configuration file. Relative paths are resolved based on the workspace root. Leave blank to auto-detect.

### `commitlint.config.extend.rules`

Commitlint rules which will be extended.

**Example configuration as JSON syntax:**
```

joshbolduc reviewed (commented):
Thanks for submitting this! I have a few suggestions.

joshbolduc commented on src/settings.ts:
```diff
@@ -1,3 +1,4 @@
+import { QualifiedConfig } from '@commitlint/types/lib/load';
```
This type is available in the root package:

```suggestion
import { QualifiedConfig } from '@commitlint/types';
```

yi-Xu-0100 commented:
@joshbolduc I try to rebase the main branch, but the package-lock.json changed after I use the `npm install`, Is there anything I need to do with this?

```powershell
vscode-commitlint [add-configuration-extend ↓5 ↑10] % npm install
npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!

> esbuild@0.11.18 postinstall D:\Documents\Creations\github\vscode-commitlint\node_modules\esbuild
> node install.js

added 485 packages from 292 contributors and audited 487 packages in 50.116s

69 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

joshbolduc commented:
Oh, I think that's from using npm 7. That was an accident, I'll look at reverting it back.

yi-Xu-0100 commented:
@joshbolduc I think that The new npm is better, I just confused for this error. Sorry for the delay reply. I will wait for your reply or any suggestion.

joshbolduc commented:
I reverted the lockfile version change in #100, you can rebase now.

npm v7 is probably better, but it's not the default in node 12 or 14, and I don't know a good way to maintain multiple versions of npm (like you can with `.nvmrc` for node itself). So I'll probably stick with v6 for the time being just to keep things simpler.

yi-Xu-0100 commented:
@joshbolduc I just update the `npm` version to v7. and when I use `npm install`, it can not identify the engine. 😅

```powershell
vscode-commitlint [add-configuration-extend ↓5 ↑8] % npm install
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'commitlint@1.2.1',
npm WARN EBADENGINE   required: { npm: '^6.0.0', vscode: '^1.52.0' },
npm WARN EBADENGINE   current: { node: 'v14.15.5', npm: '7.11.2' }
npm WARN EBADENGINE }

added 4 packages, removed 3 packages, changed 15 packages, and audited 487 packages in 6s

found 0 vulnerabilities
```

Thanks for your explanation! I will try to find a way to solve the problem.

joshbolduc reviewed (approved):
This is great! Thanks so much.
